### PR TITLE
feat(common): add createContract, createNonceManager utils

### PR DESCRIPTION
**Pull request by holic**
pulled out of #1214

### CODE REVIEW

File: packages/common/src/createContract.ts [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+} from "viem";
+import pQueue from "p-queue";
+import pRetry from "p-retry";
+import { createNonceManager } from "./createNonceManager";
+import { debug as parentDebug } from "./debug";
+
+const debug = parentDebug.extend("createContract");
+
+// copied from viem because it isn't exported
+// TODO: import from viem?
```
Comment (alvrs) [inline packages/common/src/createContract.ts:24]: should we open a viem issue for this?

File: packages/common/src/createContract.ts [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+                    // On nonce errors, reset the nonce and retry
+                    if (nonceManager.shouldResetNonce(error)) {
+                      debug("got nonce error, retrying", error);
+                      await nonceManager.resetNonce();
+                      return;
+                    }
+                    throw error;
+                  },
+                }),
+              { throwOnTimeout: true }
```
Comment (holic) [inline packages/common/src/createContract.ts:116]: we don't have a timeout configured but `queue.add` has a `Promise<T | void>` return type unless you toggle this on

File: packages/common/src/createContract.ts [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+          return async (...parameters) => {
+            const { args, options } = getFunctionParameters(parameters as any);
+
+            async function write(): Promise<Hex> {
+              if (!nonceManager.hasNonce()) {
+                await nonceManager.resetNonce();
+              }
+
+              debug("simulating write", functionName, args, options);
+              const { request } = await publicClient.simulateContract({
```
Comment (alvrs) [inline packages/common/src/createContract.ts:83]: so the `request` include the estimated gas limit correct?

**Review by alvrs — APPROVED** [review]

File: packages/common/src/createContract.ts (hunk 1 above)
Comment (holic) [reply packages/common/src/createContract.ts:24]: image

🤷‍♂️

**Pull request merged**